# ISM6644 - Python Foundamentals for Business Analytics

# Module 2 - Data Manipulation with Python

# Loading data 

In [ ]:
#Loading data from an existing library
#Load scikit-learns datasets
from sklearn import datasets

# Load the Iris dataset
iris = datasets.load_iris()

# Display some basic information about the dataset
print(iris.keys())  # Displays the keys in the dataset dictionary
print(iris['DESCR'])  # Displays a description of the dataset

In [13]:
#Import library from sklearn, extract load_iris data check keys 

from sklearn import datasets

iris = datasets.load_iris()
print(iris.keys())
print(iris['DESCR'])


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])
.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= =====

In [14]:
#import the pandas library
import pandas as pd

#create a dataframe
df = pd.DataFrame(data = iris.data, columns=iris.feature_names)

# Display the first five rows of the data
df.head()


,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2


In [ ]:
import pandas as pd

df = pd.Dataframe(data = iris.data)

In [ ]:
pip install --upgrade pandas

In [17]:
#Display last rows of data
df.tail()

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
145,6.7,3.0,5.2,2.3
146,6.3,2.5,5.0,1.9
147,6.5,3.0,5.2,2.0
148,6.2,3.4,5.4,2.3
149,5.9,3.0,5.1,1.8


SCENARIO: You work for a tech retail company analyzing laptop inventory data.
The dataset is messy (like real-world data!) and needs cleaning before analysis.

YOUR MISSION: Clean the data and discover insights about laptop pricing and specs.

In [29]:
import os
os.getcwd()

'/Users/telaclaimstech/Desktop/MSIS_Classes/python_fundamentals_Assigment_1/module2/nb'

In [172]:
#Loading a comma-separated values (CSV) data file
#Load library
import pandas as pd

#Load a file of data in table format from your computer
my_df = pd.read_excel('laptops.csv')

#view first 2 rows
my_df.head(2)

,Price ($),Screen Size (in.),RAM Memory (GB),Hard drive (GB),USB Ports,Brand,Weight (oz.),Purchased,X1,X2
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2200.0,18.4,8.0,1500.0,2.0,Apple,5.6,No,NaN,NaN


In [37]:
import pandas as pd
my_df = pd.read_excel('laptops.csv')
my_df.head(5)

,Price ($),Screen Size (in.),RAM Memory (GB),Hard drive (GB),USB Ports,Brand,Weight (oz.),Purchased,X1,X2
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2200.0,18.4,8.0,1500.0,2.0,Apple,5.6,No,NaN,NaN
2,2000.0,17.0,6.0,1000.0,2.0,HP,5.2,Yes,NaN,NaN
3,1900.0,18.4,8.0,1000.0,4.0,Toshiba,9.7,Yes,NaN,NaN
4,1900.0,13.1,6.0,1500.0,2.0,HP,5.1,Yes,NaN,NaN


# Data Understanding

In [70]:
#DataFrame Summary
my_df.info(memory_usage="deep")


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 62 entries, 0 to 61
Data columns (total 10 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Price ($)          60 non-null     float64
 1   Screen Size (in.)  61 non-null     float64
 2   RAM Memory (GB)    61 non-null     float64
 3   Hard drive (GB)    61 non-null     float64
 4   USB Ports          61 non-null     float64
 5   Brand              61 non-null     object 
 6   Weight (oz.)       61 non-null     float64
 7   Purchased          61 non-null     object 
 8   X1                 0 non-null      float64
 9   X2                 0 non-null      float64
dtypes: float64(8), object(2)
memory usage: 10.3 KB


In [41]:
#working with attributes enables you to quickly understand the structure an dproperties of your dataset without altering the data itself
#Shape provides the number of rows and columns
my_df.shape

(62, 10)

In [42]:
# Size provides the number of observations in your dataset
my_df.size

620

In [43]:
#dtype provides information about features' types
my_df.dtypes

Price ($)            float64
Screen Size (in.)    float64
RAM Memory (GB)      float64
Hard drive (GB)      float64
USB Ports            float64
Brand                 object
Weight (oz.)         float64
Purchased             object
X1                   float64
X2                   float64
dtype: object

In [44]:
#list of features
my_df.columns

Index(['Price ($)', 'Screen Size (in.)', 'RAM Memory (GB)', 'Hard drive (GB)',
       'USB Ports', 'Brand', 'Weight (oz.)', 'Purchased', 'X1', 'X2'],
      dtype='object')

In [47]:
#refers to column at index 2
my_df.columns[2]

'RAM Memory (GB)'

In [53]:
my_df['Brand'].value_counts()

Brand
HP         26
Toshiba    25
Apple      10
Name: count, dtype: int64

In [60]:
#complete the code below to count the number of laptops that have been "Purchased'
my_df['Purchased'].value_counts()

Purchased
No     32
Yes    29
Name: count, dtype: int64

In [61]:
my_df['Brand'].unique()

array([nan, 'Apple', 'HP', 'Toshiba'], dtype=object)

# Accessing Rows and Columns of the DataFrame

In [62]:
#accessing data columns
my_df['Brand']

0         NaN
1       Apple
2          HP
3     Toshiba
4          HP
       ...   
57         HP
58    Toshiba
59         HP
60         HP
61    Toshiba
Name: Brand, Length: 62, dtype: object

In [67]:
my_df.loc[:,'Price ($)']

0        NaN
1     2200.0
2     2000.0
3     1900.0
4     1900.0
       ...  
57     400.0
58     400.0
59     400.0
60     380.0
61     300.0
Name: Price ($), Length: 62, dtype: float64

In [68]:
my_df.loc[:, "Price ($)":"RAM Memory (GB)"]

,Price ($),Screen Size (in.),RAM Memory (GB)
0,NaN,NaN,NaN
1,2200.0,18.4,8.0
2,2000.0,17.0,6.0
3,1900.0,18.4,8.0
4,1900.0,13.1,6.0
...,...,...,...
57,400.0,14.0,3.0
58,400.0,13.3,2.0
59,400.0,15.6,3.0
60,380.0,14.0,3.0


In [71]:
#complete the code below and display only the first two column of Laptops' dataset
my_df.loc[:,"Price ($)":"Screen Size (in.)"]

,Price ($),Screen Size (in.)
0,NaN,NaN
1,2200.0,18.4
2,2000.0,17.0
3,1900.0,18.4
4,1900.0,13.1
...,...,...
57,400.0,14.0
58,400.0,13.3
59,400.0,15.6
60,380.0,14.0


In [87]:
#Accessing rows
my_df[1:5] #denotes the rows indexed 1 to 4


,Price ($),Screen Size (in.),RAM Memory (GB),Hard drive (GB),USB Ports,Brand,Weight (oz.),Purchased,X1,X2
1,2200.0,18.4,8.0,1500.0,2.0,Apple,5.6,No,NaN,NaN
2,2000.0,17.0,6.0,1000.0,2.0,HP,5.2,Yes,NaN,NaN
3,1900.0,18.4,8.0,1000.0,4.0,Toshiba,9.7,Yes,NaN,NaN
4,1900.0,13.1,6.0,1500.0,2.0,HP,5.1,Yes,NaN,NaN


In [75]:
#calculate the average of laptop prices
my_df.loc[:,'Price ($)'].mean()

920.75

In [80]:
#Position based in indexing
my_df.iloc[0:5,0:4]

,Price ($),Screen Size (in.),RAM Memory (GB),Hard drive (GB)
0,NaN,NaN,NaN,NaN
1,2200.0,18.4,8.0,1500.0
2,2000.0,17.0,6.0,1000.0
3,1900.0,18.4,8.0,1000.0
4,1900.0,13.1,6.0,1500.0


In [92]:
#complete the code below an show only rows 10 to 15 and columns 2 to 4 of the Laptop's data
my_df.iloc[10:16,2:5]

,RAM Memory (GB),Hard drive (GB),USB Ports
10,6.0,1000.0,4.0
11,6.0,640.0,2.0
12,6.0,1000.0,2.0
13,6.0,640.0,3.0
14,6.0,640.0,2.0
15,6.0,640.0,2.0


# Data Wrangling

In [171]:
#remove redundant rows and columns of data
new_df=my_df.iloc[1:, 0:7]
new_df.head()

,Price ($),Screen Size (in.),RAM Memory (GB),Hard drive (GB),USB Ports,Brand,Weight (oz.)
1,2200.0,18.4,8.0,1500.0,2.0,Apple,5.6
2,2000.0,17.0,6.0,1000.0,2.0,HP,5.2
3,1900.0,18.4,8.0,1000.0,4.0,Toshiba,9.7
4,1900.0,13.1,6.0,1500.0,2.0,HP,5.1
5,1800.0,15.4,8.0,1000.0,2.0,Apple,5.6


In [114]:
#rename columns of data frame
new_df.rename(columns={'Price ($)':'Price', 'Screen Size (in.)':'ScreenSize',
                        'RAM Memory (GB)':'RAM','Hard drive (GB)':'HD' },
inplace=True)

new_df.head()

,Price,ScreenSize,RAM,HD,USB Ports,Brand,Weight
1,2200.0,18.4,8.0,1500.0,2.0,Apple,5.6
2,2000.0,17.0,6.0,1000.0,2.0,HP,5.2
3,1900.0,18.4,8.0,1000.0,4.0,Toshiba,9.7
4,1900.0,13.1,6.0,1500.0,2.0,HP,5.1
5,1800.0,15.4,8.0,1000.0,2.0,Apple,5.6


In [115]:
# Replace missing value by average of the column
new_df['Price'].fillna(new_df['Price'].mean(),inplace=True)

#Check the number of missing values in a column
new_df['Price'].isna().sum()

/var/folders/pn/m2nfshws0wd37gvf2z0f6v040000gn/T/ipykernel_18836/2107826854.py:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  new_df['Price'].fillna(new_df['Price'].mean(),inplace=True)


0

In [116]:
new_df['Price'].fillna(new_df['Price'].mean, inplace=True)
new_df['Price'].isna().sum()

0

In [39]:
# Remove duplicate rows from the DataFrame
rows_before = len(new_df)
new_df = new_df.drop_duplicates()
rows_after = len(new_df)

print(f"Rows before: {rows_before}")
print(f"Rows after: {rows_after}")
print(f"Duplicates removed: {rows_before - rows_after}")

#By default, it keeps the first occurrence of the duplicate rows and drops the rest. 
#If you want to keep the last occurrence, you can specify the keep parameter:
#new_df.drop_duplicates(keep='last')

Rows before: 61
Rows after: 60
Duplicates removed: 1


In [118]:
rows_before = len(new_df)
new_df = new_df.drop_duplicates()
rows_after = len(new_df)

print(f"{rows_before}")
print(f"{rows_after}")
print(f"{rows_after - rows_before}")

60
60
0


In [45]:
# sort data
new_df_sorted = new_df.sort_values(by='Price', ascending=True)
print("\nCheapest 3 laptops:")
print(new_df_sorted.head(3))
print("\nMost expensive 3 laptops:")
print(new_df_sorted.tail(3))



Cheapest 3 laptops:
    Price  ScreenSize  RAM     HD  USB Ports    Brand  Weight (oz.)
61  300.0        15.6  3.0  250.0        2.0  Toshiba           5.5
60  380.0        14.0  3.0  320.0        3.0       HP           4.9
58  400.0        13.3  2.0  250.0        2.0  Toshiba           5.4

Most expensive 3 laptops:
    Price  ScreenSize  RAM      HD  USB Ports    Brand  Weight (oz.)
3  1900.0        18.4  8.0  1000.0        4.0  Toshiba           9.7
2  2000.0        17.0  6.0  1000.0        2.0       HP           5.2
1  2200.0        18.4  8.0  1500.0        2.0    Apple           5.6


In [47]:
# Encoding categorical data to dummy variables
df_encoded = pd.get_dummies(new_df, columns=['Brand'], dtype=int)
print(df_encoded.head())

    Price  ScreenSize  RAM      HD  USB Ports  Weight (oz.)  Brand_Apple  \
1  2200.0        18.4  8.0  1500.0        2.0           5.6            1   
2  2000.0        17.0  6.0  1000.0        2.0           5.2            0   
3  1900.0        18.4  8.0  1000.0        4.0           9.7            0   
4  1900.0        13.1  6.0  1500.0        2.0           5.1            0   
5  1800.0        15.4  8.0  1000.0        2.0           5.6            1   

   Brand_HP  Brand_Toshiba  
1         0              0  
2         1              0  
3         0              1  
4         1              0  
5         0              0  


In [51]:
#Grouping - Compare Brands
# Which Offers Best Value?

laptop_groups = new_df.groupby('Brand')

# Average specs by brand
print("\nAVERAGE SPECIFICATIONS BY BRAND")
print(laptop_groups.mean().round(2))
print("\n")

# Count laptops per brand
print("\nNUMBER OF MODELS PER BRAND")
print(laptop_groups.size())
print("\n")

# Find the best value brand (lowest average price)
brand_prices = laptop_groups['Price'].mean().sort_values()
print("\nBRANDS RANKED BY AVERAGE PRICE (Cheapest to Most Expensive)")
print(brand_prices.round(2))
print("\n")

print(f"\n INSIGHT: {brand_prices.index[0]} offers the lowest average price at ${brand_prices.iloc[0]:.2f}")
print(f"\n INSIGHT: {brand_prices.index[-1]} is the premium brand at ${brand_prices.iloc[-1]:.2f}")



AVERAGE SPECIFICATIONS BY BRAND
           Price  ScreenSize   RAM     HD  USB Ports  Weight (oz.)
Brand                                                             
Apple    1450.00       14.57  6.20  781.0       1.90          3.83
HP        823.23       15.66  4.68  621.6       3.00          5.79
Toshiba   819.40       15.49  4.56  542.8       2.52          6.06



NUMBER OF MODELS PER BRAND
Brand
Apple      10
HP         25
Toshiba    25
dtype: int64



BRANDS RANKED BY AVERAGE PRICE (Cheapest to Most Expensive)
Brand
Toshiba     819.40
HP          823.23
Apple      1450.00
Name: Price, dtype: float64



 INSIGHT: Toshiba offers the lowest average price at $819.40

 INSIGHT: Apple is the premium brand at $1450.00


Calculate which laptops offer the best specs for the price.

What is the ValueScore?
It's a custom metric that answers: "How much laptop am I getting for my money?"

How it works:

1. Numerator (top of the fraction) - Combines all the good specs:
   - `RAM` - More RAM is better (in GB, like 8 or 16)
   - `HD/100` - Hard drive size divided by 100 (so 500GB becomes 5)
     - Why divide? So HD doesn't dominate the score since it's in the hundreds
   - `ScreenSize*10` - Screen size multiplied by 10 (so 15.6 inches becomes 156)
     - Why multiply? So screen size has more weight in the final score

2. Denominator (bottom) - The price:
   - Dividing by price means: higher specs at lower prices = better value

In [53]:
# Calculate which laptops offer the best specs for the price
new_df['ValueScore'] = (new_df['RAM'] + new_df['HD']/100 + new_df['ScreenSize']*10) / new_df['Price']
best_value = new_df.nlargest(5, 'ValueScore')[['Brand', 'Price', 'RAM', 'HD', 'ScreenSize', 'ValueScore']]

print("\n")
print("Top 5 Best Value Laptops (Most specs per dollar)")
print(best_value.round(3))



Top 5 Best Value Laptops (Most specs per dollar)
      Brand  Price  RAM     HD  ScreenSize  ValueScore
61  Toshiba  300.0  3.0  250.0        15.6       0.538
59       HP  400.0  3.0  320.0        15.6       0.406
60       HP  380.0  3.0  320.0        14.0       0.385
57       HP  400.0  3.0  320.0        14.0       0.366
56  Toshiba  450.0  3.0  350.0        15.6       0.361


# In-class Practice

In [ ]:
#Load the Oil datase. 
#Identify if there are missing values and replace them by the average of the column. 
#Sort the data by 'Miles'. 
#Find the descriptive statistics by vehicle types.
#Recode Vehicle Type to dummy variales.

### Solution

In [155]:
## Access excel
import pandas as pd
data = pd.read_excel('./oil.xlsx')
data.head()

,OilChanges,Repairs,Miles,Type
0,3.0,300.0,20100,Honda
1,5.0,300.0,23200,Honda
2,2.0,500.0,19200,Honda
3,3.0,400.0,22100,Toyota
4,1.0,700.0,18400,Toyota


In [156]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   OilChanges  13 non-null     float64
 1   Repairs     13 non-null     float64
 2   Miles       14 non-null     int64  
 3   Type        14 non-null     object 
dtypes: float64(2), int64(1), object(1)
memory usage: 580.0+ bytes


In [157]:
## Transform to Dataframe
df = pd.DataFrame(data=data)
df

,OilChanges,Repairs,Miles,Type
0,3.0,300.0,20100,Honda
1,5.0,300.0,23200,Honda
2,2.0,500.0,19200,Honda
3,3.0,400.0,22100,Toyota
4,1.0,700.0,18400,Toyota
5,4.0,420.0,23400,Nissan
6,6.0,100.0,17900,Nissan
7,4.0,290.0,19900,Toyota
8,3.0,475.0,20100,Toyota
9,2.0,620.0,24100,Toyota


In [158]:
## Fill NA missing values with mean
count_na = df.isna().sum()
print(count_na)

df['OilChanges'].fillna(df['OilChanges'].mean(), inplace=True)
df['OilChanges'].isna().sum()

df['Repairs'].fillna(df['Repairs'].mean(), inplace=True)
df['OilChanges'].isna().sum()

df

OilChanges    1
Repairs       1
Miles         0
Type          0
dtype: int64


/var/folders/pn/m2nfshws0wd37gvf2z0f6v040000gn/T/ipykernel_18836/2014829592.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['OilChanges'].fillna(df['OilChanges'].mean(), inplace=True)
/var/folders/pn/m2nfshws0wd37gvf2z0f6v040000gn/T/ipykernel_18836/2014829592.py:8: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setti

,OilChanges,Repairs,Miles,Type
0,3.000000,300.000000,20100,Honda
1,5.000000,300.000000,23200,Honda
2,2.000000,500.000000,19200,Honda
3,3.000000,400.000000,22100,Toyota
4,1.000000,700.000000,18400,Toyota
5,4.000000,420.000000,23400,Nissan
6,6.000000,100.000000,17900,Nissan
7,4.000000,290.000000,19900,Toyota
8,3.000000,475.000000,20100,Toyota
9,2.000000,620.000000,24100,Toyota


In [166]:
## Sorted by Miles ascending
df_sorted = df.sort_values(by='Miles',ascending=True)
df_sorted.head(15)

,OilChanges,Repairs,Miles,Type
6,6.000000,100.000000,17900,Nissan
10,4.461538,600.000000,18200,Honda
4,1.000000,700.000000,18400,Toyota
2,2.000000,500.000000,19200,Honda
11,10.000000,381.153846,19600,Honda
13,8.000000,50.000000,19700,Nissan
7,4.000000,290.000000,19900,Toyota
0,3.000000,300.000000,20100,Honda
8,3.000000,475.000000,20100,Toyota
12,7.000000,200.000000,20800,Honda


In [160]:
# Get statistics for specific column Type
df_stat = df.groupby('Type').describe()
df_stat

OilChanges                                                         \
            count     mean       std  min       25%       50%  75%   max   
Type                                                                       
Honda         6.0  5.24359  2.895995  2.0  3.365385  4.730769  6.5  10.0   
Nissan        3.0  6.00000  2.000000  4.0  5.000000  6.000000  7.0   8.0   
Toyota        5.0  2.60000  1.140175  1.0  2.000000  3.000000  3.0   4.0   

       Repairs              ...                    Miles                \
         count        mean  ...         75%    max count          mean   
Type                        ...                                          
Honda      6.0  380.192308  ...  470.288462  600.0   6.0  20183.333333   
Nissan     3.0  190.000000  ...  260.000000  420.0   3.0  20333.333333   
Toyota     5.0  497.000000  ...  620.000000  700.0   5.0  20920.000000   

                                                                  
                std      min      25%      50%      75%      max  
Type                                                              
Honda   1716.294458  18200.0  19300.0  19850.0  20625.0  23200.0  
Nissan  2804.163571  17900.0  18800.0  19700.0  21550.0  23400.0  
Toyota  2211.786608  18400.0  19900.0  20100.0  22100.0  24100.0  

[3 rows x 24 columns]

In [167]:
## Use pd.get_dummies to trasnform table specific column into dummy 0 and 1
df_encoded = pd.get_dummies(df_sorted,columns=['Type'], dtype=int)
print(df_encoded)

    OilChanges     Repairs  Miles  Type_Honda  Type_Nissan  Type_Toyota
6     6.000000  100.000000  17900           0            1            0
10    4.461538  600.000000  18200           1            0            0
4     1.000000  700.000000  18400           0            0            1
2     2.000000  500.000000  19200           1            0            0
11   10.000000  381.153846  19600           1            0            0
13    8.000000   50.000000  19700           0            1            0
7     4.000000  290.000000  19900           0            0            1
0     3.000000  300.000000  20100           1            0            0
8     3.000000  475.000000  20100           0            0            1
12    7.000000  200.000000  20800           1            0            0
3     3.000000  400.000000  22100           0            0            1
1     5.000000  300.000000  23200           1            0            0
5     4.000000  420.000000  23400           0            1      

### Summary

In [ ]:
## Access variables and Use iloc and loc 

In [ ]:
## Data Wrangling:bran

In [ ]:
## ValueScore: new_df['ValueScore'] = (new_df['RAM] + new_df['HD']/100 + new_df['ScreenSize']*10 / new_df['Price])
## nlargest value scores: new_df.nlargest(5, 'ValueScore')[['Brand','Price','RAM','HD','ScreenSize','ValueScore']

In [ ]:
## Value score over total price, nlargest to get top of those
## Create new variable ValueScore
## Then use nlargest to get highest values for that formula and their columns
## print a round of those values

new_df['ValueScore'] = (new_df['RAM'] + new_df['HD']/100 + new_df['ScreenSize']*10) / new_df['Price']
best_value = new_df.nlargest(10, 'ValueScore')[['Brand','Price','RAM','HD','ScreenSize','ValueScore']]

print(best_value.round(3))

In [ ]:
## Practice Questions: Accessing Variables with iloc and loc

## 1. How would you select ALL rows but only the 'Brand' column using .loc?
## 2. How would you select the 'Price ($)' column through 'RAM Memory (GB)' column (inclusive) using .loc?
## 3. How would you select rows 10 through 15 and columns 2 through 4 using .iloc?
## 4. What is the difference between .loc and .iloc?
## 5. How would you access the value at row index 0 and the first 5 columns using .iloc?

## Practice Questions: Data Wrangling

## 6. How would you rename the columns of new_df in place using a dictionary?
## 7. How would you fill missing values in the 'Price' column with that column's average, in place?
## 8. How would you check how many missing values remain in the 'Price' column?
## 9. How would you count the number of rows in new_df before removing duplicates?
## 10. How would you remove duplicate rows from new_df?
## 11. How would you count the number of rows in new_df after removing duplicates?
## 12. How would you sort new_df by 'Price' in ascending order?
## 13. How would you encode the 'Brand' column into dummy variables using pd.get_dummies?

## Practice Questions: Grouping and Comparing Brands

## 14. How would you group new_df by 'Brand'?
## 15. How would you find the average of each numeric column per brand, rounded to 2 decimals?
## 16. How would you find the number of laptop models per brand?
## 17. How would you calculate the average price per brand, sorted from cheapest to most expensive?
## 18. How would you print an insight identifying the brand with the lowest average price and its price?
## 19. How would you print an insight identifying the brand with the highest average price and its price?

## Practice Questions: ValueScore

## 20. How would you create a new 'ValueScore' column combining RAM, HD, and ScreenSize as specs, divided by Price?
## 21. Why is HD divided by 100 and ScreenSize multiplied by 10 in the formula?
## 22. How would you retrieve the top 5 laptops with the highest ValueScore, showing only Brand, Price, RAM, HD, ScreenSize, and ValueScore?
## 23. How would you print those results rounded to 3 decimal places?

In [34]:
import pandas as pd
data = pd.read_excel('./laptops.csv')
df = pd.DataFrame(data)
df

,Price ($),Screen Size (in.),RAM Memory (GB),Hard drive (GB),USB Ports,Brand,Weight (oz.),Purchased,X1,X2
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2200.0,18.4,8.0,1500.0,2.0,Apple,5.6,No,NaN,NaN
2,2000.0,17.0,6.0,1000.0,2.0,HP,5.2,Yes,NaN,NaN
3,1900.0,18.4,8.0,1000.0,4.0,Toshiba,9.7,Yes,NaN,NaN
4,1900.0,13.1,6.0,1500.0,2.0,HP,5.1,Yes,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...
57,400.0,14.0,3.0,320.0,3.0,HP,4.9,No,NaN,NaN
58,400.0,13.3,2.0,250.0,2.0,Toshiba,5.4,No,NaN,NaN
59,400.0,15.6,3.0,320.0,3.0,HP,5.6,Yes,NaN,NaN
60,380.0,14.0,3.0,320.0,3.0,HP,4.9,No,NaN,NaN


In [7]:
df.loc[:,'Brand']

0         NaN
1       Apple
2          HP
3     Toshiba
4          HP
       ...   
57         HP
58    Toshiba
59         HP
60         HP
61    Toshiba
Name: Brand, Length: 62, dtype: object

In [8]:
df.loc[:,'Price ($)':'RAM Memory (GB)']

,Price ($),Screen Size (in.),RAM Memory (GB)
0,NaN,NaN,NaN
1,2200.0,18.4,8.0
2,2000.0,17.0,6.0
3,1900.0,18.4,8.0
4,1900.0,13.1,6.0
...,...,...,...
57,400.0,14.0,3.0
58,400.0,13.3,2.0
59,400.0,15.6,3.0
60,380.0,14.0,3.0


In [10]:
df.iloc[10:15,2:5]

,RAM Memory (GB),Hard drive (GB),USB Ports
10,6.0,1000.0,4.0
11,6.0,640.0,2.0
12,6.0,1000.0,2.0
13,6.0,640.0,3.0
14,6.0,640.0,2.0


In [ ]:
#. loc uses name and .iloc index 

In [12]:
df.iloc[:1,:5]

,Price ($),Screen Size (in.),RAM Memory (GB),Hard drive (GB),USB Ports
0,NaN,NaN,NaN,NaN,NaN


In [60]:
df.rename(columns={'Price ($)':'Price','Screen Size (in.)':'Screen_Size', 'RAM Memory (GB)':'RAM', 'Hard drive (GB)': 'GB', 'Weight (oz.)': 'Weight'}, inplace=True)
df

,Price,Screen_Size,RAM,GB,USB Ports,Brand,Weight,Purchased,X1,X2
0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2200.0,18.4,8.0,1500.0,2.0,Apple,5.6,No,NaN,NaN
2,2000.0,17.0,6.0,1000.0,2.0,HP,5.2,Yes,NaN,NaN
3,1900.0,18.4,8.0,1000.0,4.0,Toshiba,9.7,Yes,NaN,NaN
4,1900.0,13.1,6.0,1500.0,2.0,HP,5.1,Yes,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...
57,400.0,14.0,3.0,320.0,3.0,HP,4.9,No,NaN,NaN
58,400.0,13.3,2.0,250.0,2.0,Toshiba,5.4,No,NaN,NaN
59,400.0,15.6,3.0,320.0,3.0,HP,5.6,Yes,NaN,NaN
60,380.0,14.0,3.0,320.0,3.0,HP,4.9,No,NaN,NaN


In [21]:
na_count = df.isna().sum()
print(na_count)

Price           2
Screen_Size     1
RAM             1
GB              1
USB Ports       1
Brand           1
Weight          1
Purchased       1
X1             62
X2             62
dtype: int64


In [23]:
df['Price'].fillna(df['Price'].mean(), inplace=True)
df['Price'].isna().sum()

/var/folders/pn/m2nfshws0wd37gvf2z0f6v040000gn/T/ipykernel_29369/2513820645.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Price'].fillna(df['Price'].mean(), inplace=True)


0

In [25]:
rows_before = len(df)
print(rows_before)

62


In [27]:
df_clean = df.drop_duplicates()
rows_after = len(df_clean)
print(rows_after)

61


In [28]:
df_clean_sorted = df_clean.sort_values(by='Price', ascending=True)
df_clean_sorted

,Price,Screen_Size,RAM,GB,USB Ports,Brand,Weight,Purchased,X1,X2
61,300.0,15.6,3.0,250.0,2.0,Toshiba,5.5,No,NaN,NaN
60,380.0,14.0,3.0,320.0,3.0,HP,4.9,No,NaN,NaN
58,400.0,13.3,2.0,250.0,2.0,Toshiba,5.4,No,NaN,NaN
57,400.0,14.0,3.0,320.0,3.0,HP,4.9,No,NaN,NaN
59,400.0,15.6,3.0,320.0,3.0,HP,5.6,Yes,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...
5,1800.0,15.4,8.0,1000.0,2.0,Apple,5.6,No,NaN,NaN
4,1900.0,13.1,6.0,1500.0,2.0,HP,5.1,Yes,NaN,NaN
3,1900.0,18.4,8.0,1000.0,4.0,Toshiba,9.7,Yes,NaN,NaN
2,2000.0,17.0,6.0,1000.0,2.0,HP,5.2,Yes,NaN,NaN


In [37]:
df_dummies = pd.get_dummies(df_clean_sorted,columns=['Brand'], dtype=int)
print(df_dummies)

     Price  Screen_Size  RAM      GB  USB Ports  Weight Purchased  X1  X2  \
61   300.0         15.6  3.0   250.0        2.0     5.5        No NaN NaN   
60   380.0         14.0  3.0   320.0        3.0     4.9        No NaN NaN   
58   400.0         13.3  2.0   250.0        2.0     5.4        No NaN NaN   
57   400.0         14.0  3.0   320.0        3.0     4.9        No NaN NaN   
59   400.0         15.6  3.0   320.0        3.0     5.6       Yes NaN NaN   
..     ...          ...  ...     ...        ...     ...       ...  ..  ..   
5   1800.0         15.4  8.0  1000.0        2.0     5.6        No NaN NaN   
4   1900.0         13.1  6.0  1500.0        2.0     5.1       Yes NaN NaN   
3   1900.0         18.4  8.0  1000.0        4.0     9.7       Yes NaN NaN   
2   2000.0         17.0  6.0  1000.0        2.0     5.2       Yes NaN NaN   
1   2200.0         18.4  8.0  1500.0        2.0     5.6        No NaN NaN   

    Brand_Apple  Brand_HP  Brand_Toshiba  
61            0         0       

In [51]:
brand_group = df_clean_sorted.groupby('Brand')
brand_group.mean(numeric_only=True).round(2)

,Price,Screen_Size,RAM,GB,USB Ports,Weight,X1,X2
Brand,,,,,,,,
Apple,1450.00,14.57,6.20,781.0,1.90,3.83,NaN,NaN
HP,823.23,15.66,4.68,621.6,3.00,5.79,NaN,NaN
Toshiba,819.40,15.49,4.56,542.8,2.52,6.06,NaN,NaN


In [53]:
brand_group.size()

Brand
Apple      10
HP         25
Toshiba    25
dtype: int64

In [54]:
brand_prices = brand_group['Price'].mean().sort_values()
brand_prices.round(2)

Brand
Toshiba     819.40
HP          823.23
Apple      1450.00
Name: Price, dtype: float64

In [58]:
print(f'Lowest Brand: {brand_prices.index[0]}, Lowest Price: {brand_prices.iloc[0]:.2f}')
print(f'Highest Brand: {brand_prices.index[-1]}, Highest Price: {brand_prices.iloc[-1]:.2f}')

Lowest Brand: Toshiba, Lowest Price: 819.40
Highest Brand: Apple, Highest Price: 1450.00


In [68]:
df['ValueScore'] = (df['RAM'] + df['GB']/100 + df['Screen_Size']*10) / df['Price']

In [69]:
best_values = df.nlargest(5,'ValueScore')[['Brand','Price','RAM','GB','Screen_Size','ValueScore']]
print(best_values.round(3))

      Brand  Price  RAM     GB  Screen_Size  ValueScore
61  Toshiba  300.0  3.0  250.0         15.6       0.538
59       HP  400.0  3.0  320.0         15.6       0.406
60       HP  380.0  3.0  320.0         14.0       0.385
57       HP  400.0  3.0  320.0         14.0       0.366
56  Toshiba  450.0  3.0  350.0         15.6       0.361
